feature: Added option for custom function for edge orientation in Exp…

Green-Kedia opened the pull request:
…ertInLoop algorithm

### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #1953

### List of changes to the codebase in this pull request
- Changes only in `pgmpy/estimators/expert.py` to add the flexibility for user to pass a custom function that provides the orientation of a queried edge
- Also add the possible edges from u->v or v->u in `blacklisted_edges` if the `orientation_fn` gives None so that no edge in future is added.
- Updated the docstrings accordingly

Green-Kedia committed:
feature: Added option for custom function for edge orientation in ExpertInLoop algorithm

Green-Kedia committed:
Added tests for the edits in expert.py to include the option of custom function for deciding edge orientation

Green-Kedia committed:
Added tests for the edits in expert.py to include the option of custom function for deciding edge orientation

Green-Kedia committed:
Merge branch 'dev' into feature/custom-orientation_fn-expertinloop

ankurankan commented on pgmpy/estimators/expert.py:
```diff
@@ -89,53 +88,91 @@ def estimate(
-        use_llm: bool
-            Whether to use a Large Language Model for edge orientation. If
-            False, prompts the user to specify the direction between the edges.
+        orientation_fn: callable, (default: pgmpy.utils.llm_pairwise_orient)
```
Should mention here that any additional kwargs argument will be passed to this method as well.

ankurankan commented on pgmpy/estimators/expert.py:
```diff
@@ -188,37 +225,58 @@ def estimate(
             edge_direction = None
-            if use_llm:
+
+            if orientations:
```
Could you please add a comment here with "Step 3.2: ..."

ankurankan replied:
I think would help with readability if you could write the overall if-else flow in the comment.

ankurankan commented on pgmpy/estimators/expert.py:
```diff
@@ -14,6 +14,7 @@
     def __init__(self, data=None, **kwargs):
         super(ExpertInLoop, self).__init__(data=data, **kwargs)
+        self.orientations_from_fn = set([])
         self.orientations_llm = set([])
```
These two are used for caching. Right? Can we get rid of one?

ankurankan reviewed (changes requested):
@Green-Kedia Thanks for the PR. I think it mostly looks good; I have added a few minor comments.

Green-Kedia committed:
modified llm_pairwise_orient and manual_pairwise_orient to return None incase of no edge. Also adjusted tests for absence of orientations_llm cache

Green-Kedia commented:
@ankurankan I have done some changes 
- As without `orientations_llm` a test was failing so I adjusted the tests according to whichever cache attribute is present.
- Also as you asked, I have updated the llm and manual pairwise orient functions to accept None as an option. 
- I have also updated the doctorings as you asked

any more suggestions?

ankurankan commented on pgmpy/estimators/expert.py:
```diff
@@ -14,7 +14,7 @@
     def __init__(self, data=None, **kwargs):
         super(ExpertInLoop, self).__init__(data=data, **kwargs)
-        self.orientations_llm = set([])
+        self.orientations_from_fn = set([])
```
I think a more readable name could be something like `orientation_cache`

ankurankan reviewed (changes requested):
@Green-Kedia Sorry, but can you revert the changes for returning None. I had another look at the algorithm, and I think we will need to make some modifications to the algorithm to allow None. I also think the algorithm is more general without the None option (it usually harder to tell if a causal relationship doesn't exist), and should automatically fix any incorrect edges. Sorry about the confusion.

Green-Kedia committed:
reverted changes, and now it does not accepts None as a return option for custom function, also fixed a few names and docstrings

Green-Kedia commented:
@ankurankan No problem at all. I’ve reverted the changes and updated a few docstrings accordingly. Apologies for the obvious complications in the attribute names, I’ve corrected them now.
Let me know if there’s anything else that needs to be changed.

ankurankan reviewed (approved):
@Green-Kedia Looks good. Thanks again for the PR.
